### DISCUSSION TITLE
Connect multiple RPI clients to same GDS instance

### DISCUSSION THREAD

**Discussion by prithvisathiyawild** [nasa/fprime#1278] [Q&A]
## Our setup

We are looking to have multiple Raspberry PI FSW connected to the same GDS instance. The idea is to be able to separately command multiple spacecrafts and view aggregated telemetry all at once, primarily for ground testing purposes. This does require modifications to the framing/deframing of the packets to correctly encode the target satellite into the data message itself, which we have already done so by dynamically setting the appropriate `FpFramerDeframer.START_TOKEN`.

## Problem

We currently have a single instance of GDS running, binded at `--ip-port=55000`, which works as expected when different satellites are connected to it **one at a time**. However, when connecting 2 RPI's to the same port, only the first successfully connects, while the second client socket request is never accepted by the `TcpHandler` within `adapters/ip.py`. 

I have tried increasing the `IpHandler.MAX_CLIENT_BACKLOG` to >1, but that doesn't allow for concurrent connections. It seems the only way to enable this, is to edit the `ip.py` file to start a new thread for each client connection, allowing for multiple `socket.accept()` calls.

Has anyone else also attempted to refactor the GDS for multi-client support? If so, is the approach described above (with threading) the way to go?

**Comment by Emil808** [comment]
I am also working on a project with the same goal in mind. I have tested similar to having two instances of the same deployment run and connect to the same gds(host/port). I observed that only the first one to connect will get accepted, while the second will be held waiting in the backlog. Then when the first deployment disconnects, the second gets accepted. 

The here is the repo for the fprime-gds: https://github.com/fprime-community/fprime-gds.

A threaded TCP server seems the way to go. I though, haven't made any progress yet as I'm still trying to understand how the gds connection pipeline works and I'm not as experienced with TCP programming. 

Something to consider later is that the fprime-gds relies on a dictionary to know what cmds, events, and tlm there is in the deployment. If each device is truly identical, then you'll need to bring up the `START_TOKEN` up with the cmd/tlm/event, up to display, rather than leaving it at the framer/deframer.

I asked something similar earlier here: https://github.com/nasa/fprime/discussions/1107
Though, at the time, I was looking at the TCPClient and TCPServer component implementations for deployments. And actually, we want to modify the gds.

**Comment by prithvisathiyawild** [comment]
Weird, i posted a reply here a couple days ago, but it seems like it was deleted. Will summarize my solution here again shortly.

**Comment by prithvisathiyawild** [comment]
So yes, i was already able to figure this out by refactoring `IpAdapter` class to handle multi threaded client sockets. And correct, the threaded TCP server is the appropriate approach. 

Here's a quick overview of the changes I made, to make the whole things work. Keep in mind, most of the changes within `ip.py` are a combination between the `ThreadedTCPServer/ThreadedTCPRequestHandler` in `tcpserver.py` and the example posted in this [documentation](https://docs.python.org/3/library/socketserver.html#asynchronous-mixins).

1. `ip.py`
    - Instead of the default `TcpHandler`, create a `ThreadedTCPHandler` and override the `handle` method to handle the incoming client sockets. U will need to keep track of these sockets in so some list/dict to refer to them later
    - Pass the handler object to a new `ThreadedTCPServer` class. Then spawn of this server in its own thread along with the `ThreadedTCPServer.serve_forever` method, so the server is non-blocking. (i.e. set `self.thtcp = threading.Thread(target=self.server.serve_forever)`
    - Modify the `close`, `write` and `read` methods of the adapter to correctly write to & read from the set of sockets. This will depend on the target satellite u choose to interact with, so ensure the sockets are appropriately referenced. 
2. `framing.py`
    - There will no longer be a static token, since there are multiple satellites. Modify the `START_TOKEN` dynamically, and assign the correct token. Keep in mind, any incoming messages from different satellites will have different tokens. 
3. `updown.py`
    - Depending on ur implementation, u will need to parse or embed the satellite into the message prior to framing or post deframing. 
4. There a bunch of other places that will need to be refactored upstream, all the way to the actual GUI, such that the satellite id flows E2E. But that is highly dependent on ur implementation and how/where u want to embed that information. 

If u have questions on specific components of the refactor, happy to post examples here.
